### DISCUSSION TITLE
brew external command --help

### DISCUSSION THREAD

**Discussion by tddschn** [Homebrew/discussions#3326] [Tap maintenance and Homebrew development] [answered]
I made a python script named `brew-myapp` and can be run via `brew myapp <args> <opts>`, 
the script itself provides `--help`, but it's gone if not run with `brew-myapp` (with the dash).

From the [doc](https://docs.brew.sh/External-Commands#providing---help):
> Providing --help
All internal and external Homebrew commands can provide styled --help output by using Homebrew’s argument parser, as seen in the brew services command; or by including lines starting with #: (a comment then : character in both Bash and Ruby), as seen in the header of update.sh, which is printed with brew update --help.

The --help of my script is automatically and dynamically generated, it doesn't have a `#: ` comment section and I don't know how to make it work with Homebrew’s argument parser.

How can I make `--help` work? Thank you.

**Comment by gromgit** [comment, marked as answer]
If I'm reading the Homebrew code correctly, it intercepts any argument in the form of:
* `-h`
* `--help`
* `--usage`
* `-?`

and invokes internal help code that looks for a `#:` comment section in your script.

So if you absolutely must generate your help text dynamically, or otherwise have your script handle `--help`, I think you only have one option: **Rename `--help` to something else, like `help`.**

For instance:
```
$ cat cmd/brew-test-tcl
#!/usr/bin/env tclsh
switch [lindex $::argv 0] {
  help {
    puts "Hi! You've asked for help with {[lrange $::argv 1 end]}."
  }
  default {
    puts "ARGS: $::argv"
  }
}

$ brew test-tcl help me obi-wan
Hi! You've asked for help with {me obi-wan}.

$ brew test-tcl no-droids here
ARGS: no-droids here
```

**Comment by tddschn** [comment]
I changed my help option from `-h, --help` to `-help`, \
`brew myapp -help` and `brew myapp <args> <opts>` works as expected,  \
except very slow compared to `brew-myapp -help` and `brew-myapp <args> <opts>`.

How do I make brew just let my script run? Thank you.

**Reply by gromgit** [reply]
> How do I make brew just let my script run?

In short, you can't. As long as you're calling your script via `brew myapp`, you're really running Homebrew itself (`brew`), which does a whole bunch of environment sanitization and internal prep, before looking for your `brew-myapp` script in your local taps. All this will happen every time you run `brew myapp`.

**Comment by tddschn** [comment]
I thought brew just need to inject some `HOMEBREW_*` env vars - why does brew need to do so many things for an external command?

`git` for example doesn't do anything for a custom sub command, at least not anything that could noticably impact performance.

**Reply by gromgit** [reply]
Not a Homebrew maintainer, but I can think of several reasons:
1. Git is (largely) a single native executable, while Homebrew is a collection of Ruby scripts, with a smattering of Bash, so there's a fairly large disparity in startup time for the base command.
2. Git literally _internalizes_ almost all internal subcommands (read: single native executable), while there's no physical distinction between internal and external Homebrew subcommands--they're all individual scripts.
3. Git assumes you know what you're doing, and that your environment is set up correctly. Homebrew tries very hard to make sure it's being run in a sane environment with necessary prerequisites, and recommends corrective actions where necessary (see `$(brew --repo)/Library/Homebrew/brew.*` for the gory logic).
